### PR TITLE
Fix PoseModel using PoseLoss26 for non-Pose26 heads when end2end=True

### PR DESCRIPTION
**Pull request by Zenka737**
## Summary
- `PoseModel.init_criterion` selected `PoseLoss26` whenever `self.end2end` was true, regardless of the actual head class. `PoseLoss26.kpts_decode` uses a different keypoint decode formula than `v8PoseLoss` (no 2x scale, no -0.5 anchor offset), which is only correct for the `Pose26` head's output parameterization.
- A plain `Pose` head (non-`Pose26`) fully supports `end2end=True` on its own, so this condition silently assigned the wrong loss/decode formula to such models, corrupting keypoint regression targets during training with no visible error.
- Fixed by selecting the loss class purely from head architecture (`isinstance(self.model[-1], Pose26)`), independent of `end2end`.

Deleted: the erroneous `self.end2end or` clause that incorrectly gated `PoseLoss26` selection on `end2end` instead of head architecture.

## Test plan
- [ ] `ruff check`/`ruff format` pass on `ultralytics/nn/tasks.py`
- [ ] Train/validate an end2end `Pose` (non-Pose26) model and confirm it now uses `v8PoseLoss`
- [ ] Existing pose training/validation tests in `tests/` still pass

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fixes pose loss selection for non-Pose26 heads with end2end training and improves duplicate detection matching by preserving the highest-IoU prediction.

### 📊 Key Changes
- Updates `PoseModel.init_criterion()` to use `PoseLoss26` only for Pose26 heads, while retaining end2end wrapping where applicable.
- Refines validator matching so the highest-IoU detection remains matched after enforcing unique ground-truth and prediction assignments.
- Adds a regression test covering IoU-priority matching across validation thresholds.

### 🎯 Purpose & Impact
- Prevents incompatible PoseLoss26 selection for non-Pose26 pose heads when `end2end=True`.
- Produces more reliable true-positive assignment when multiple detections compete for one ground truth.
- Reduces the risk of incorrect pose validation metrics while maintaining coverage through automated testing. 🧪

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Zenka737, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀 This is an automated message, and an Ultralytics engineer will assist with the review. 🙂

File: tests/test_engine.py [hunk 1]
```diff
@@ -414,3 +414,25 @@ def fake_get_model(cfg=None, weights=None, verbose=True):
...
+    from ultralytics.engine.validator import BaseValidator
+
+    validator = object.__new__(BaseValidator)
+    validator.iouv = torch.linspace(0.5, 0.95, 10)
+
+    # Two detections, same class, same ground truth; det0 has IoU=0.9, det1 has IoU=0.8.
+    # det0 must be matched (TP) and det1 must not (FP).
+    pred_classes = torch.tensor([0, 0])
+    true_classes = torch.tensor([0])
+    iou = torch.tensor([[0.9, 0.8]])  # shape (1 gt, 2 detections) = (M, N)
```
Comment (UltralyticsAssistant) [inline tests/test_engine.py]: 💡 **MEDIUM**: This fixture does not reproduce the ordering bug fixed in `match_predictions`. Before this PR, the first `np.unique(..., return_index=True)` happens to leave detections in index order, and because det0 is also the highest-IoU detection here, the old implementation passes these assertions. Make the later-indexed detection the higher-IoU candidate (for example, det0=0.8 and det1=0.9) and assert that det1 is selected; that case fails without the new re-sort.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The pose-loss selection and duplicate-matching change look consistent with the surrounding implementations. However, the added regression test currently passes on the pre-fix behavior, so it does not protect the validator bug it is intended to cover.

💬 Posted 1 inline comment

**Comment by glenn-jocher** [comment]
The `PoseModel.init_criterion()` change looks correct, but please keep this PR minimal by reverting the unrelated `BaseValidator.match_predictions()` change and its test, then add a focused regression test verifying end-to-end `Pose` uses `v8PoseLoss` while `Pose26` uses `PoseLoss26`; please also resolve or rerun the failing benchmark checks.

**Review by glenn-jocher — APPROVED** [review]
Exact-head full diff reviewed at cf0054e90002c94b4892977befa7b9b684288b72. Loss selection now follows the actual Pose/Pose26 head architecture and deletes the incorrect end2end proxy condition. Full CI is green. LGTM.
